components/brave_wallet/browser/buy_and_sell_service.cc

```diff
@@ -0,0 +1,566 @@
+  }
+
+  return url;
+}
+
+void BuyAndSellService::GetServiceProviders(
+    const std::string& countries,
+    const std::string& fiat_currencies,
+    const std::string& crypto_currencies,
+    const std::string& service_providers,
+    const std::string& payment_method_types,
+    const std::string& statuses,
+    GetServiceProvidersCallback callback) {
+  auto internal_callback =
+      base::BindOnce(&BuyAndSellService::OnGetServiceProviders,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+
+  api_request_helper_->Request(
+      "GET",
+      GetServiceProviderURL(countries, fiat_currencies, crypto_currencies,
+                            service_providers, payment_method_types, statuses),
+      "", "", std::move(internal_callback), MakeMeldApiHeaders(),
+      {.auto_retry_on_network_change = true, .enable_cache = true});
+}
+
+void BuyAndSellService::OnGetServiceProviders(
+    GetServiceProvidersCallback callback,
+    APIRequestResult api_request_result) const {
+  if (!api_request_result.Is2XXResponseCode()) {
+    std::move(callback).Run({},
+                            std::vector<std::string>{"INTERNAL_SERVICE_ERROR"});
+    return;
+  }
+
+  if (std::vector<std::string> errors;
+      ParseMeldErrorResponse(api_request_result.value_body(), &errors)) {
+    std::move(callback).Run({}, errors);
+    return;
+  }
+
+  std::vector<mojom::ServiceProviderPtr> service_providers;
+  if (!ParseServiceProviders(api_request_result.value_body(),
+                             &service_providers)) {
+    std::move(callback).Run({}, std::vector<std::string>{"PARSING_ERROR"});
+    return;
+  }
+  std::move(callback).Run(std::move(service_providers), std::nullopt);
+}
+
+void BuyAndSellService::GetCryptoQuotes(
+    const std::string& country,
+    const std::string& source_currency_code,
+    const std::string& destination_currency_code,
+    const double source_amount,
+    const std::string& account,
+    GetCryptoQuotesCallback callback) {
+  base::Value::Dict payload;
+  AddKeyIfNotEmpty(&payload, "countryCode", country);
+  AddKeyIfNotEmpty(&payload, "sourceCurrencyCode", source_currency_code);
+  AddKeyIfNotEmpty(&payload, "sourceAmount",
+                   base::NumberToString(source_amount));
+  AddKeyIfNotEmpty(&payload, "destinationCurrencyCode",
+                   destination_currency_code);
+  AddKeyIfNotEmpty(&payload, "walletAddress", account);
+
+  const std::string json_payload = GetJSON(payload);
+
+  auto url = GURL(base::StringPrintf("%s/payments/crypto/quote",
+                                     GetMeldAssetRatioBaseURL().c_str()));
+  auto internal_callback =
+      base::BindOnce(&BuyAndSellService::OnGetCryptoQuotes,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+
+  api_request_helper_->Request(
+      "POST", url, json_payload, "application/json",
+      std::move(internal_callback), MakeMeldApiHeaders(),
```

Comment:
You should use a conversion callback here to sanitise the JSON and convert the numeric types to strings.